File: brainfuck.rb
def brain_luck(code, input)
  code_arr, input_arr = code.split(''), input.split('').map(&:ord)
  mac = Machine.new code_arr, input_arr
  mac.run
end

class Machine

  def initialize instructions, input
    # Initialize data and pointer
    @data = Array.new(300, 0)
    @data_ptr = 0

    # Initialize instructions and pointer
    @inst = instructions
    @inst_ptr = 0

    @input = input
    @output = ''

    # index the '[' and corresponding ']' locations
    index_braces
  end

  def run
    loop { break if interpret.nil? }
    @output
  end


  private

  def index_braces
    stack = []
    @left_braces, @right_braces = [], []
    @inst.each_with_index do |x, i|
      stack << i if x == '['
      if x == ']'
        @left_braces << stack.pop
        @right_braces << i
      end
    end
  end

  def interpret
    case @inst[@inst_ptr]
    when '>'
      @data_ptr += 1
    when '<'
      @data_ptr -= 1
    when '+'
      @data[@data_ptr] = (datum + 1) % 256
    when '-'
      @data[@data_ptr] = (datum - 1) % 256
    when '.'
      @output += datum.chr
    when ','
      @data[@data_ptr] = @input.shift
    when '['
      lbi = @left_braces.index @inst_ptr
      @inst_ptr = @right_braces[lbi] if datum.zero?
    when ']'
      rbi = @right_braces.index @inst_ptr
      @inst_ptr = @left_braces[rbi] unless datum.zero?
    else
      return nil
    end
    @inst_ptr += 1
  end

  def datum
    @data[@data_ptr]
  end
end


puts brain_luck(',+[-.,+]', 'Codewars' + 255.chr) == 'Codewars'
puts brain_luck(',[.[-],]', 'Codewars' + 0.chr) == 'Codewars'
puts brain_luck(',>,<[>[->+>+<<]>>[-<<+>>]<<<-]>>.', 8.chr + 9.chr) == 72.chr

File: brainfuck.js
function brainLuck(code, input){
  return Machine.run(code, input);
}

var Machine = (function(my) {
  var data, dataPointer, instructionPointer, instructions, input, output, rightBraces, leftBraces
  
  function initialize(code, myInput){
    data = Array.apply(null, new Array(300)).map(Number.prototype.valueOf, 0);
    dataPointer = 150;
    instructionPointer = 0;
    output = '';
    rightBraces = [];
    leftBraces = [];
    instructions = code.split('');
    input = myInput.split('').map(
      function(a){
        return a.charCodeAt(0);
      });
      
    parseBraces();
  }
  
  my.run = function(code, myInput) {
    initialize(code, myInput);
    while (instructionPointer < instructions.length && instructionPointer >= 0){
      interpreter();  
    }
    return output;
  }
  
  function parseBraces(){
    var tempLeft = [];
    for (var i = 0; i < instructions.length; i++){
      if (instructions[i] == '['){
        tempLeft.push(i);
      } else if (instructions[i] == ']'){
        rightBraces.push(i);
        leftBraces.push(tempLeft.pop());
      }
    }
  }
  
  function interpreter(){
    switch (instructions[instructionPointer]){
      case '>':
        dataPointer++;
        break;
      case '<':
        dataPointer--;
        break;
      case '+':
        data[dataPointer] = (data[dataPointer] + 1) % 256;
        break;
      case '-':
        data[dataPointer] = (data[dataPointer] - 1) % 256;
        break;
      case '.':
        output += String.fromCharCode(data[dataPointer]);
        break;
      case ',':
        data[dataPointer] = input.shift();
        break;
      case '[':
        if (data[dataPointer] === 0){
          instructionPointer = rightBraces[leftBraces.indexOf(instructionPointer)] + 1;
          return void 0;
        }
        break;
      case ']':
        if (data[dataPointer] !== 0){
          instructionPointer = leftBraces[rightBraces.indexOf(instructionPointer)] + 1;
          return void 0;
        }
        break;
    }
    instructionPointer++;
    return void 0;
  }
  
  return my;
})(Machine || {});
